File: frameworks/rails/app/controllers/benchmark_controller.rb
# frozen_string_literal: true

require 'zlib'
require 'pg'

class BenchmarkController < ActionController::API
  mattr_accessor :dataset, :dataset_large, :database_path, :static_files

  DATA_DIR = ENV.fetch('DATA_DIR', '/data')
  dataset_path = File.join(DATA_DIR, 'dataset.json')
  dataset_large_path = File.join(DATA_DIR, 'dataset-large.json')
  database_path = File.join(DATA_DIR, 'benchmark.db')
  static_dir = File.join(DATA_DIR, 'static')

  if File.exist?(dataset_path)
    self.dataset = JSON.parse(File.read(dataset_path)).freeze
  end

  if File.exist?(dataset_large_path)
    raw = JSON.parse(File.read(dataset_large_path))
    items = raw.map { |d| d.merge('total' => (d['price'] * d['quantity'] * 100).round / 100.0) }
    self.dataset_large = JSON.generate({ 'items' => items, 'count' => items.length }).freeze
  end

  if File.exist?(database_path)
    self.database_path = database_path
  end

  # Load static files into memory
  MIME_TYPES = {
    '.css'   => 'text/css',
    '.js'    => 'application/javascript',
    '.html'  => 'text/html',
    '.woff2' => 'font/woff2',
    '.svg'   => 'image/svg+xml',
    '.webp'  => 'image/webp',
    '.json'  => 'application/json'
  }.freeze

  self.static_files = {}
  if File.exist?(static_dir)
    Dir.foreach(static_dir) do |name|
      next if name == '.' || name == '..'
      path = File.join(static_dir, name)
      next unless File.file?(path)
      ext = File.extname(name)
      ct = MIME_TYPES.fetch(ext, 'application/octet-stream')
      self.static_files[name] = { path: path, content_type: ct }
    end
  end
  self.static_files.freeze

  DB_QUERY = 'SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN ? AND ? LIMIT 50'.freeze
  PG_QUERY = 'SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN $1 AND $2 LIMIT 50'.freeze

  def pipeline
    render plain: 'ok'
  end

  def baseline11
    total = 0
    request.query_parameters.each_value do |v|
      total += v.to_i
    end
    if request.post?
      body_str = request.body.read
      total += body_str.to_i
    end
    render plain: total.to_s
  end

  def baseline2
    total = 0
    request.query_parameters.each_value do |v|
      total += v.to_i
    end
    render plain: total.to_s
  end

  def json_endpoint
    if dataset
      items = dataset.map { |d| d.merge('total' => (d['price'] * d['quantity'] * 100).round / 100.0) }
      body = JSON.generate({ 'items' => items, 'count' => items.length })
      response.headers['content-type'] = 'application/json'
      render plain: body
    else
      head 500
    end
  end

  def compression
    accept_encodings = request.headers['Accept-Encoding'].split(',').map(&:strip)
    if accept_encodings.include? 'gzip'
      sio = StringIO.new
      gz = Zlib::GzipWriter.new(sio, 1)
      gz.write(self.class.dataset_large)
      gz.close
      response.headers['content-type'] = 'application/json'
      response.headers['content-encoding'] = 'gzip'
      send_data sio.string, disposition: :inline
    else
      render json: self.class.dataset_large
    end
  end

  def db
    min_val = (params[:min] || 10).to_i
    max_val = (params[:max] || 50).to_i

    rows = self.class.get_db_statement&.with do |statement|
      statement.execute([min_val, max_val])
    end || []

    items = rows.map do |r|
      {
        'id' => r['id'], 'name' => r['name'], 'category' => r['category'],
        'price' => r['price'], 'quantity' => r['quantity'], 'active' => r['active'] == 1,
        'tags' => JSON.parse(r['tags']),
        'rating' => { 'score' => r['rating_score'], 'count' => r['rating_count'] }
      }
    end
    render json: { items: items, count: items.length }
  end

  def async_db
    min_val = (params[:min] || 10).to_i
    max_val = (params[:max] || 50).to_i

    rows = self.class.get_async_db&.with do |connection|
      connection.exec_prepared('select', [min_val, max_val])
    end || []

    items = rows.map do |r|
      {
        'id' => r['id'].to_i, 'name' => r['name'], 'category' => r['category'],
        'price' => r['price'].to_f, 'quantity' => r['quantity'].to_i,
        'active' => r['active'] == 't',
        'tags' => JSON.parse(r['tags']),
        'rating' => { 'score' => r['rating_score'].to_f, 'count' => r['rating_count'].to_i }
      }
    end
    render json: { items: items, count: items.length }
  end

  def static_file
    filename = params[:filename]
    if static_file = static_files[filename]
      send_file static_file[:path], type: static_file[:content_type], disposition: :inline
    else
      head 404
    end
  end

  def upload
    size = 0
    buf = request.body
    while (chunk = buf.read(65536))
      size += chunk.bytesize
    end
    render plain: size.to_s
  end

  def not_found
    head 404
  end

  private

  def self.get_db_statement
    @db_statement ||= begin
      return unless database_path
      max_connections = ENV.fetch('RAILS_MAX_THREADS', 4).to_i
      ConnectionPool.new(size: max_connections, timeout: 5) do
        db = SQLite3::Database.new(database_path, readonly: true)
        db.execute('PRAGMA mmap_size=268435456')
        db.results_as_hash = true
        db.prepare(DB_QUERY)
      end
    end
  end

  def self.get_async_db
    @async_db ||= begin
      return unless ENV['DATABASE_URL'].present?
      max_connections = ENV.fetch('RAILS_MAX_THREADS', 4).to_i
      ConnectionPool.new(size: max_connections, timeout: 5) do
        db = PG.connect(ENV['DATABASE_URL'])
        db.prepare('select', PG_QUERY)
        db
      end
    end
  end
end

File: frameworks/spring-jvm/src/main/java/com/httparena/BenchmarkController.java
package com.httparena;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PostConstruct;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.nio.file.Files;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.Properties;

@RestController
public class BenchmarkController {

    private final ObjectMapper mapper = new ObjectMapper();
    private List<Map<String, Object>> dataset;
    private byte[] largeJsonResponse;
    private boolean dbAvailable = false;
    private final Map<String, byte[]> staticFiles = new ConcurrentHashMap<>();
    private static final String DB_QUERY = "SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN ? AND ? LIMIT 50";
    private static final String PG_QUERY = "SELECT id, name, category, price, quantity, active, tags, rating_score, rating_count FROM items WHERE price BETWEEN ? AND ? LIMIT 50";
    private static final ThreadLocal<Connection> tlConn = new ThreadLocal<>();
    private HikariDataSource pgPool;
    private static final Map<String, String> MIME_TYPES = Map.of(
        ".css", "text/css", ".js", "application/javascript", ".html", "text/html",
        ".woff2", "font/woff2", ".svg", "image/svg+xml", ".webp", "image/webp", ".json", "application/json"
    );

    @PostConstruct
    public void init() throws IOException {
        String path = System.getenv("DATASET_PATH");
        if (path == null) path = "/data/dataset.json";
        File f = new File(path);
        if (f.exists()) {
            dataset = mapper.readValue(f, new TypeReference<>() {});
        }
        File largef = new File("/data/dataset-large.json");
        if (largef.exists()) {
            List<Map<String, Object>> largeDataset = mapper.readValue(largef, new TypeReference<>() {});
            List<Map<String, Object>> largeItems = new ArrayList<>(largeDataset.size());
            for (Map<String, Object> item : largeDataset) {
                Map<String, Object> processed = new LinkedHashMap<>(item);
                double price = ((Number) item.get("price")).doubleValue();
                int quantity = ((Number) item.get("quantity")).intValue();
                processed.put("total", Math.round(price * quantity * 100.0) / 100.0);
                largeItems.add(processed);
            }
            largeJsonResponse = mapper.writeValueAsBytes(Map.of("items", largeItems, "count", largeItems.size()));
        }
        dbAvailable = new File("/data/benchmark.db").exists();
        // Initialize PostgreSQL connection pool from DATABASE_URL
        String dbUrl = System.getenv("DATABASE_URL");
        if (dbUrl != null && !dbUrl.isEmpty()) {
            try {
                URI uri = new URI(dbUrl.replace("postgres://", "postgresql://"));
                String host = uri.getHost();
                int port = uri.getPort() > 0 ? uri.getPort() : 5432;
                String database = uri.getPath().substring(1);
                String[] userInfo = uri.getUserInfo().split(":");
                HikariConfig config = new HikariConfig();
                config.setDriverClassName("org.postgresql.Driver");
                config.setJdbcUrl("jdbc:postgresql://" + host + ":" + port + "/" + database);
                config.setUsername(userInfo[0]);
                config.setPassword(userInfo.length > 1 ? userInfo[1] : "");
                config.setMaximumPoolSize(64);
                config.setMinimumIdle(16);
                pgPool = new HikariDataSource(config);
            } catch (Exception ignored) {}
        }
        File staticDir = new File("/data/static");
        if (staticDir.isDirectory()) {
            File[] files = staticDir.listFiles();
            if (files != null) {
                for (File sf : files) {
                    if (sf.isFile()) {
                        try {
                            staticFiles.put(sf.getName(), Files.readAllBytes(sf.toPath()));
                        } catch (IOException ignored) {}
                    }
                }
            }
        }
    }

    @GetMapping(value = "/pipeline", produces = MediaType.TEXT_PLAIN_VALUE)
    public String pipeline() {
        return "ok";
    }

    @GetMapping(value = "/baseline11", produces = MediaType.TEXT_PLAIN_VALUE)
    public String baselineGet(@RequestParam Map<String, String> params) {
        return String.valueOf(sumParams(params));
    }

    @PostMapping(value = "/baseline11", produces = MediaType.TEXT_PLAIN_VALUE)
    public String baselinePost(@RequestParam Map<String, String> params, @RequestBody String body) {
        int sum = sumParams(params);
        try {
            sum += Integer.parseInt(body.trim());
        } catch (NumberFormatException ignored) {}
        return String.valueOf(sum);
    }

    @GetMapping(value = "/baseline2", produces = MediaType.TEXT_PLAIN_VALUE)
    public String baseline2(@RequestParam Map<String, String> params) {
        return String.valueOf(sumParams(params));
    }

    @GetMapping(value = "/compression", produces = MediaType.APPLICATION_JSON_VALUE)
    public org.springframework.http.ResponseEntity<byte[]> compression() throws IOException {
        java.io.ByteArrayOutputStream baos = new java.io.ByteArrayOutputStream();
        java.util.zip.GZIPOutputStream gzip = new java.util.zip.GZIPOutputStream(baos) {{
            def.setLevel(java.util.zip.Deflater.BEST_SPEED);
        }};
        gzip.write(largeJsonResponse);
        gzip.close();
        return org.springframework.http.ResponseEntity.ok()
            .header("Content-Type", "application/json")
            .header("Content-Encoding", "gzip")
            .body(baos.toByteArray());
    }

    @GetMapping(value = "/json", produces = MediaType.APPLICATION_JSON_VALUE)
    public Map<String, Object> json() {
        List<Map<String, Object>> items = new ArrayList<>(dataset.size());
        for (Map<String, Object> item : dataset) {
            Map<String, Object> processed = new LinkedHashMap<>(item);
            double price = ((Number) item.get("price")).doubleValue();
            int quantity = ((Number) item.get("quantity")).intValue();
            processed.put("total", Math.round(price * quantity * 100.0) / 100.0);
            items.add(processed);
        }
        return Map.of("items", items, "count", items.size());
    }

    @PostMapping(value = "/upload", produces = MediaType.TEXT_PLAIN_VALUE)
    public String upload(InputStream body) throws IOException {
        byte[] buf = new byte[65536];
        long total = 0;
        int n;
        while ((n = body.read(buf)) != -1) {
            total += n;
        }
        return String.valueOf(total);
    }

    @GetMapping(value = "/db", produces = MediaType.APPLICATION_JSON_VALUE)
    public byte[] db(@RequestParam(defaultValue = "10") double min, @RequestParam(defaultValue = "50") double max) throws IOException {
        if (!dbAvailable) {
            return "{\"items\":[],\"count\":0}".getBytes();
        }
        Connection conn = getDbConnection();
        List<Map<String, Object>> items = new ArrayList<>();
        try {
            PreparedStatement stmt = conn.prepareStatement(DB_QUERY);
            stmt.setDouble(1, min);
            stmt.setDouble(2, max);
            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                Map<String, Object> item = new LinkedHashMap<>();
                item.put("id", rs.getLong("id"));
                item.put("name", rs.getString("name"));
                item.put("category", rs.getString("category"));
                item.put("price", rs.getDouble("price"));
                item.put("quantity", rs.getInt("quantity"));
                item.put("active", rs.getInt("active") == 1);
                item.put("tags", mapper.readValue(rs.getString("tags"), new TypeReference<List<String>>() {}));
                item.put("rating", Map.of("score", rs.getDouble("rating_score"), "count", rs.getInt("rating_count")));
                items.add(item);
            }
            rs.close();
            stmt.close();
        } catch (SQLException e) {
            return "{\"items\":[],\"count\":0}".getBytes();
        }
        return mapper.writeValueAsBytes(Map.of("items", items, "count", items.size()));
    }

    private Connection getDbConnection() {
        Connection conn = tlConn.get();
        if (conn == null) {
            try {
                Properties props = new Properties();
                props.setProperty("open_mode", "1");  // SQLITE_OPEN_READONLY
                conn = DriverManager.getConnection("jdbc:sqlite:/data/benchmark.db", props);
                conn.createStatement().execute("PRAGMA mmap_size=268435456");
                tlConn.set(conn);
            } catch (SQLException e) {
                throw new RuntimeException(e);
            }
        }
        return conn;
    }

    @GetMapping(value = "/async-db", produces = MediaType.APPLICATION_JSON_VALUE)
    public byte[] asyncDb(@RequestParam(defaultValue = "10") double min, @RequestParam(defaultValue = "50") double max) throws IOException {
        if (pgPool == null) {
            return "{\"items\":[],\"count\":0}".getBytes();
        }
        try (Connection conn = pgPool.getConnection()) {
            PreparedStatement stmt = conn.prepareStatement(PG_QUERY);
            stmt.setDouble(1, min);
            stmt.setDouble(2, max);
            ResultSet rs = stmt.executeQuery();
            List<Map<String, Object>> items = new ArrayList<>();
            while (rs.next()) {
                Map<String, Object> item = new LinkedHashMap<>();
                item.put("id", rs.getLong("id"));
                item.put("name", rs.getString("name"));
                item.put("category", rs.getString("category"));
                item.put("price", rs.getDouble("price"));
                item.put("quantity", rs.getInt("quantity"));
                item.put("active", rs.getBoolean("active"));
                item.put("tags", mapper.readValue(rs.getString("tags"), new TypeReference<List<String>>() {}));
                item.put("rating", Map.of("score", rs.getDouble("rating_score"), "count", rs.getInt("rating_count")));
                items.add(item);
            }
            rs.close();
            stmt.close();
            return mapper.writeValueAsBytes(Map.of("items", items, "count", items.size()));
        } catch (SQLException e) {
            return "{\"items\":[],\"count\":0}".getBytes();
        }
    }

    @GetMapping("/static/{filename}")
    public org.springframework.http.ResponseEntity<byte[]> staticFile(@PathVariable String filename) {
        byte[] data = staticFiles.get(filename);
        if (data == null) {
            return org.springframework.http.ResponseEntity.notFound().build();
        }
        int dot = filename.lastIndexOf('.');
        String ext = dot >= 0 ? filename.substring(dot) : "";
        String ct = MIME_TYPES.getOrDefault(ext, "application/octet-stream");
        return org.springframework.http.ResponseEntity.ok()
            .header("Content-Type", ct)
            .body(data);
    }

    private int sumParams(Map<String, String> params) {
        int sum = 0;
        for (String v : params.values()) {
            try {
                sum += Integer.parseInt(v);
            } catch (NumberFormatException ignored) {}
        }
        return sum;
    }
}
